odoo/addons/base/models/ir_async.py

```diff
@@ -0,0 +1,242 @@
+import logging
+import re
+import traceback
+import threading
+import odoo
+import odoo.exceptions
+from odoo import models, fields, api
+from odoo.http import serialize_exception
+
+_logger = logging.getLogger(__name__)
+BASE_VERSION = odoo.modules.load_information_from_description_file('base')['version']
+
+
+class IrAsync(models.Model):
+    _name = 'ir.async'
+    _description = 'Asynchrone jobs'
+
+    name = fields.Char(default="Unamed background task")
+    state = fields.Selection([
+        ('created', 'Created'),
+        ('succeeded', 'Succeeded'),
+        ('failed', 'Failed'),
+        ('done', 'Done'),
+    ])
+    res_model = fields.Char()
+    res_ids = fields.Text()
+    method = fields.Char()
+    args = fields.Text()
+    kwargs = fields.Text()
+    user_id = fields.Many2one('res.users', string='Scheduler User', ondelete='cascade')
+    context = fields.Text()
+    super_user = fields.Boolean()
+    traceback = fields.Text()
+    payload = fields.Text(default=False)
+
+    def call(self, target, args=None, kwargs=None):
+        """
+        Thread/Process-like API to create asynchronous jobs.
+
+        Ensure the method is called in an independant worker using a
+        copy of the current environnement (user, context, recordset).
+
+        The job return value is stored has JSON in the payload field,
+        web-notification are possible via bus/models/ir_async:IrAsync.call
+        """
+        recs = getattr(target, '__self__', None)
+        if recs is None or not isinstance(recs, models.BaseModel):
+            raise TypeError("You can only create an async task on a recordset")
+        model = recs.__class__
+
+        job = self.sudo().create({
+            'state': 'created',
+            'res_model': model._name,
+            'res_ids': json.dumps(recs._ids),
+            'method': target.__name__,
+            'args': json.dumps(args or []),
+            'kwargs': json.dumps(kwargs or {}),
+            'user_id': int(recs.env.uid),
+            'context': json.dumps(recs.env.context),
+            'super_user': recs.env.su,
+            'traceback': self._merge_traceback(traceback.format_stack()[:-1]),
+        })
+
+        self._cr.after("commit", self._notifydb)
+
+        return job
+
+    def _notifydb(self):
+        """ Notify the workers a job has been enqueued """
+        with odoo.sql_db.db_connect('postgres').cursor() as cr:
+            cr.execute('NOTIFY odoo_async, %s', (self.env.cr.dbname,))
+        _logger.debug("Async task commited. Workers notified.")
+
+    @classmethod
+    def _process_jobs(cls, dbname):
+        """ Process all jobs of the selected database """
+        db = odoo.sql_db.db_connect(dbname)
+        threading.current_thread().dbname = dbname
+
+        # make sure we can process this db
+        with db.cursor() as cr:
+            cr.execute("""
+                SELECT latest_version
+                FROM ir_module_module
+                WHERE name=%s""", ('base',))
```

Comment:
Bad indentation and trivial query using 4 lines
```suggestion
            query = "SELECT latest_version FROM ir_module_module WHERE name=%s"
            cr.execute(query, ['base'])
```